home-manager: Bind mount directories instead of symlinking them

talyz opened the pull request:
Use `bindfs` to create bind mounts for directories instead of symlinking them. This should be less problematic for many applications, since bind mounts are much more transparent.

One service per bind mount is created to keep track of it. In addition, a bind mount is briefly created in the activation script, before `home-manager` switches to the new profile generation, and torn down again at the end of the activation script. This is done to ensure that the links `home-manager` creates end up in the right place; the user services aren't started until either the end of the activation script or the user session begins, and then it's too late.

lovesegfault commented:
Will this need to incorporate the changes in #3?

cole-h commented on home-manager.nix:
```diff
@@ -43,43 +43,132 @@ in
+              if ! mount | grep "${mountPoint}"; then
+                  ${pkgs.bindfs}/bin/bindfs -f --no-allow-other "${targetDir}" "${mountPoint}"
+              else
+                  echo "There is already an active mount at or below ${mountPoint}!" >&2
```
Should this exit with an error? Otherwise, I'm envisioning a scenario where the `ExecStop` is executed on a mountpoint already "owned" by something else (and thus unmounting something one might not want it to).

talyz replied:
Ah, yes, you're right! I forgot add an `exit`.

talyz commented on home-manager.nix:
```diff
@@ -6,7 +6,7 @@ let
   persistentStoragePaths = attrNames cfg;
 
-  inherit (pkgs.callPackage ./lib.nix { }) splitPath dirListToPath concatPaths;
+  inherit (pkgs.callPackage ./lib.nix { }) splitPath dirListToPath concatPaths sanitizeName;
```
No reason other than less typing, really. I'm good with either, so we can change it if you want :)

cole-h replied:
If you or someone else can verify that it works properly when using `import`, I'll send a separate PR changing this (no need for mostly-unrelated changes). Depending on merge order, one of us will just have to rebase :P

talyz commented on home-manager.nix:
```diff
@@ -43,43 +43,132 @@ in
+                dir;
             targetDir = concatPaths [ persistentStoragePath dir ];
+            mountPoint = concatPaths [ config.home.homeDirectory mountDir ];
+            systemctl = "XDG_RUNTIME_DIR=\${XDG_RUNTIME_DIR:-/run/user/$(id -u)} ${config.systemd.user.systemctlPath}";
```
Not strictly, no, but it makes the code easier to read where `systemctl` is invoked, I think. The same goes for many things in these modules, which are put in a `let` and given a name, despite only being called from one place; it makes for less heavily indented code with clearer function invocations when complex definitions aren't done inline.

talyz replied:
I guess an alternative would be to put this on the previous line with a `\` at the end, but it still feels messier to me..

cole-h replied:
I think it's fine as-is, then. I can't decide what I dislike more -- single-use variables, or unnecessary multilines when you can just use a variable... :D

talyz commented:
> Will this need to incorporate the changes in #3?

Hm, I'm not sure. The user will always be right, since the units are running as the correct user to start with, but I guess the permissions set on parent directories won't be propagated. I'm not sure if there's a use case where this matters - if there is, at least a part of the script would have to be replicated. `bindfs` is not able to somehow give enhanced access to anything in persistent storage, though, so in that sense the restrictions set still apply.

talyz commented:
I've pushed some changes to the systemd service now: mainly that it's not restarted on change, since that doesn't work very well if there are programs currently reading from the mount. Also, it now unmounts lazily, so that at least the directory tree is cleaned up, in case the filesystem is busy - previously, this would result in a disconnected endpoint, since the fuse process quit without cleaning up properly.

One fragility I ran into, which I'm not sure exactly how to fix in a nice way, is if the activation script fails when the bind mounts are still active. This results in some or all bind mounts still being alive when the script exits and can result in the same disconnected endpoint issue when the fuse processes are killed by systemd.

lovesegfault commented:
Alright, I tried to use this today but had a number of issues, some of which required this patch to be fixed:
```patch
diff --git a/home-manager.nix b/home-manager.nix
index 37d60c4..db7831c 100644
--- a/home-manager.nix
+++ b/home-manager.nix
@@ -78,7 +78,7 @@ in
             name = "bindMount-${sanitizeName targetDir}";
             startScript = pkgs.writeShellScript name ''
               set -eu
-              if ! mount | grep "${mountPoint}"; then
+              if ! ${pkgs.utillinux}/bin/mount | ${pkgs.gnugrep}/bin/grep "${mountPoint}"; then
                   ${pkgs.bindfs}/bin/bindfs -f --no-allow-other "${targetDir}" "${mountPoint}"
               else
                   echo "There is already an active mount at or below ${mountPoint}!" >&2
@@ -151,8 +151,8 @@ in
             if [[ ! -e "${mountPoint}" ]]; then
                 mkdir -p "${mountPoint}"
             fi
-            if mount | grep "${mountPoint}"; then
-                if mount | grep "${mountPoint}" | grep "${targetDir}"; then
+            if ${pkgs.utillinux}/bin/mount | ${pkgs.gnugrep}/bin/grep "${mountPoint}"; then
+                if ${pkgs.utillinux}/bin/mount | ${pkgs.gnugrep}/bin/grep "${mountPoint}" | ${pkgs.gnugrep}/bin/grep "${targetDir}"; then
                     mountedPaths["${mountPoint}"]=0
                 else
                     # The target directory changed, so we need to remount
```

Regardless, after applying the above I saw some weirdness:
```log
Jun 16 23:41:02 foucault systemd[1]: Starting Home Manager environment for bemeurer...
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]: Activating home-manager configuration for bemeurer
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]: Starting home manager activation
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]: Sanity checking Nix
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]: Sanity checking oldGenNum and oldGenPath
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]: This is a live run
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]: Using Nix version:
Jun 16 23:41:02 foucault hm-activate-bemeurer[5923]: Using Nix version:
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]: Activation variables:
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]:   oldGenNum=44
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]:   oldGenPath=/nix/store/zav06dj6l5z0aqd890vnjmdh7p678ym0-home-manager-generation
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]:   newGenPath=/nix/store/zav06dj6l5z0aqd890vnjmdh7p678ym0-home-manager-generation
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]:   newGenNum=45
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]:   newGenGcPath=/nix/var/nix/gcroots/per-user/bemeurer/current-home
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]:   genProfilePath=/nix/var/nix/profiles/per-user/bemeurer/home-manager
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]: Activating checkFilesChanged
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]: Activating checkLinkTargets
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]: Activating declareMountedPathsArray
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]: Activating writeBoundary
Jun 16 23:41:02 foucault hm-activate-bemeurer[5888]: Activating createAndMount-state-home-bemeurer
Jun 16 23:41:02 foucault systemd[1]: home-manager-bemeurer.service: Main process exited, code=exited, status=1/FAILURE
Jun 16 23:41:02 foucault systemd[1]: home-manager-bemeurer.service: Failed with result 'exit-code'.
Jun 16 23:41:02 foucault systemd[1]: Failed to start Home Manager environment for bemeurer.
```

```log
-- Logs begin at Tue 2020-06-16 23:38:10 PDT, end at Tue 2020-06-16 23:41:02 PDT. --
Jun 16 23:38:20 foucault systemd[3465]: Started Bind mount /state/home/bemeurer/.local/share/lollypop at /home/bemeurer/.local/share/lollypop.
Jun 16 23:38:20 foucault systemd[3465]: bindMount-state-home-bemeurer-local-share-lollypop.service: Main process exited, code=exited, status=1/FAILURE
Jun 16 23:38:20 foucault systemd[3465]: bindMount-state-home-bemeurer-local-share-lollypop.service: Failed with result 'exit-code'.
```

```log
Jun 16 23:46:29 foucault systemd[3379]: Started Bind mount /state/home/bemeurer/.cache/lollypop at /home/bemeurer/.cache/lollypop.
Jun 16 23:46:29 foucault mhl234ig0whs41zvsbkxah9r3py59van-bindMount-state-home-bemeurer-cache-lollypop[3393]: /state/home/bemeurer/.cache/lollypop on /home/bemeurer/.cache/lollypop type fuse (rw,nosuid,nodev,relatime,user_id=8888,group_id=100,default_permissions)
Jun 16 23:46:29 foucault mhl234ig0whs41zvsbkxah9r3py59van-bindMount-state-home-bemeurer-cache-lollypop[3387]: There is already an active mount at or below /home/bemeurer/.cache/lollypop!
Jun 16 23:46:29 foucault systemd[3379]: bindMount-state-home-bemeurer-cache-lollypop.service: Main process exited, code=exited, status=1/FAILURE
Jun 16 23:46:29 foucault systemd[3379]: bindMount-state-home-bemeurer-cache-lollypop.service: Failed with result 'exit-code'.
```

```log
Jun 16 23:48:47 foucault hm-activate-bemeurer[5754]: Activating declareMountedPathsArray
Jun 16 23:48:47 foucault hm-activate-bemeurer[5754]: Activating writeBoundary
Jun 16 23:48:47 foucault hm-activate-bemeurer[5754]: Activating createAndMount-state-home-bemeurer
Jun 16 23:48:47 foucault hm-activate-bemeurer[6111]: mkdir: cannot stat ‘/home/bemeurer/.cache/lollypop’: Transport endpoint is not connected
```

I'll add that I had other units failing as well, I just picked this one to inspect.

energizah commented:
May I suggest adding a note to the docs that the names of user mounts will be visible to other users (e.g. via the `mount` command), since this could be a data privacy issue in some circumstances.

talyz commented:
@lovesegfault I've fixed the path to `mount` (`grep` is already in `PATH`) and pushed a few changes which should make this more robust, mainly:

- Unmount all bind mounts in the activation script on error
- Try to unmount nicely a few times before resorting to a lazy unmount

I don't know why it failed for you to begin with, since it's not indicated by your log output, and this likely doesn't solve that, but

- Can you check the complete output in the journal? Frequently, the error isn't visible in the `systemctl status` output, but is when looking at it through `journalctl`.
- This is almost guaranteed to require a reboot when switching from the old symlink system, so if you only tried switching, try a reboot too.

@energizah Good catch! I've added a note in the `README.org` file.

johnae commented:
Hey there @talyz and others! Thanks for this effort! I've been using a somewhat different approach based on btrfs and snapshots previously but that didn't include /home at all. Since then I found this project and tried to use it but home-manager + impermanence didn't work all that well, possibly because of ordering of operations or something. But I'll get to the point - why bother with home-manager integration at all when it seems it would be more robust to just use binds mounts and links for both "system" files and "user" files (eg. owned by some user). Am I missing something here perhaps?

I've played around with something along those lines (eg. no home-manager integration at all) based on code from this repo here: https://github.com/johnae/world/blob/master/modules/state.nix and it works pretty well for me. Would that be of interest to integrate into this repo or are you taking a different route here? Sorry if I hijacked this pull request a bit.

talyz commented:
@johnae I talked a little about my take on it here: https://github.com/nix-community/impermanence/issues/1#issuecomment-640484933. Leveraging home-manager gives us a few things for free - file and directory ownership is handled automatically, clean-up is done when switching generations and the API is cleaner. If you had issues getting the home-manager module working, could you elaborate on that in a new issue?

talyz commented:
@lovesegfault I've been using this for over a month now without major issues. Could you try again? I fixed a few things since you tried it last.

johnae commented:
@talyz This is one thing I had an issue with earlier: https://github.com/rycee/home-manager/blob/a49ce0e9edcd717ce6e66041aabd656c02903904/modules/programs/mbsync.nix#L161 - it creates a directory as part of activation, before impermanence kicks in so activation then failed. I also had other issues such as home-manager complaining it won't write files outside $HOME (because of linking ofc). I guess this pull request is meant to solve the last one at least...

johnae commented:
And of course, you've got a point when it comes to cleanup which we'd get automatically via home-manager.

talyz commented:
@johnae It should solve both problems. Just to make sure, I just pushed a commit that puts the bind mounts before the write boundary, which should guarantee that anything written by home-manager respects persistent storage. I'd appreciate if you could try it out and see if it solves your issue.

lovesegfault commented:
@talyz a side-effect of this is that it seems like `root` can't see the bind mounted dirs:
```
~
viins ❯ echo $USER
bemeurer
~
viins ❯ ls
Permissions Size User     Date Modified Name
drwx------     - bemeurer  4 Jul 23:28  documents/
drwx------     - bemeurer  4 Jun  2:24  opt/
drwx------     - bemeurer 29 Jun 19:57  src/
drwx------     - bemeurer 19 Jul 14:28  tmp/
~
viins ❯ sudo su

[root@foucault:/home/bemeurer]# echo $USER
root

[root@foucault:/home/bemeurer]# ls
ls: cannot access 'tmp': Permission denied
ls: cannot access 'src': Permission denied
ls: cannot access 'opt': Permission denied
ls: cannot access 'documents': Permission denied
documents  opt  src  tmp

[root@foucault:/home/bemeurer]#
```

johnae commented:
I guess another issue is that any file watchers based on notify won't work as bindfs is fuse based.

talyz commented:
@lovesegfault Yes, that's a side-effect of using fuse without the `allow_other` option, to my knowledge. I haven't experienced any major issues with this, since `root` can still access the original path, though.

@johnae Yes, that also seems to be a limitation of this. I haven't run into any issues stemming from this yet and I suppose the temporary "solution" is to watch the original path where possible. This is what I had to do for dropbox, since it doesn't support fuse file systems..

johnae commented:
While this is an elegant solution in some ways, for me it would mean implementing work arounds for file watchers in shared repos which would be a bit of a mess. I think I'll keep using my own take on this despite the (different) drawbacks of that approach.

talyz commented:
@johnae That's fine :) I'm not opposed to improving the NixOS module and give people the choice to use it or this - I just think there's quite a bit of work to be done in order for it to be viable. Hopefully the stuff needed for fuse to work with fuse makes it into the kernel, too.

johnae commented:
@talyz Hope I didn't sound negative or anything. I think this is really great work. So what you're saying is that it might be interesting to merge something into the NixOS module for /home?

talyz commented:
@lovesegfault The empty directory list bug should now be fixed and I've squashed it down to three commits. I think it's ready to be merged, but it would be great if you could try it out one last time first. :)

@johnae No, it's cool - at the moment all solutions have their drawbacks and I understand if this doesn't work for you. Yes, I mean that if we get a complete solution for managing `/home` via the NixOS module, I'm not opposed to having that available alongside this one.

talyz commented:
I'll merge this now, since I've been using it without major issues for more than a month at this point. If there are major issues with it, we can just revert it.
